buildpacks/ruby/src/ruby_version.rs

```diff
@@ -0,0 +1,208 @@
+use crate::gemfile_lock::{self, EngineVersion, LockfileRuby};
+use commons::env_command::{CommandError, EnvCommand};
+use indoc::formatdoc;
+use libcnb::build::BuildContext;
+use libcnb::data::layer_name;
+use libcnb::layer_env::Scope;
+use libcnb::Env;
+use libherokubuildpack::log as user;
+use serde::Deserialize;
+use serde::Serialize;
+use std::fmt::Display;
+
+use crate::layers::RubyInstallLayer;
+use crate::RubyBuildpack;
+use crate::RubyBuildpackError;
+
+/// Used when no version is found in the Gemfile.lock
+/// Default version is not sticky
+const DEFAULT_RUBY_VERSION: &str = "3.1.3";
+
+/// Env key to disable Ruby installation. When set this tells the Ruby
+/// buildpack that we want to skip Ruby installation and instead rely on
+/// "system ruby" basically whatever is on the PATH that responds to
+/// `which ruby` this may come from a buildpack or the host OS.
+const HEROKU_USE_SYSTEM_RUBY: &str = "HEROKU_USE_SYSTEM_RUBY";
+
+/// Return a `Ruby` version based on Gemfile.lock contents and `HEROKU_USE_SYSTEM_RUBY`
+///
+/// ## Errors
+///
+/// - `HEROKU_USE_SYSTEM_RUBY` is set, but `ruby -v` errors
+/// - Ruby engine version is specified, without a Ruby version
+pub(crate) fn from_lockfile(lockfile: &str, env: &Env) -> Result<Ruby, RubyVersionError> {
+    if let Some(value) = env.get("HEROKU_USE_SYSTEM_RUBY") {
+        user::log_warning(
+            "Using System Ruby",
+            formatdoc! {"
+                Found environment variable {HEROKU_USE_SYSTEM_RUBY}={value:?}. When enabled, you are
```

Comment:
As we're brainstorming an interface. One thought/question here is debug-ability. 

I'm calling `ruby -v` and erroring if I didn't find a Ruby version on disk. To help debug if that fails I can use https://github.com/schneems/which_problem (perfect use case for that logic TBH). However there's no visibility into where the user was expecting the Ruby version to come from. I think they would want to know:

- Did a user set `HEROKU_USE_SYSTEM_RUBY` or was it a buildpack (or host OS)? 
  - We can detect if it came from platform env or in current env
- If it came from buildpack/host-os how can we forward that location to the end user?
  - One idea would be to have another env var like `HEROKU_USE_SYSTEM_RUBY_REASON` and make it mandatory when not set via platform/user env. Presumably if it's set via another buildpack it would be something like `HEROKU_USE_SYSTEM_RUBY_REASON="core-os-ruby buildpack is installing Ruby"`
  - We could just use the value of `HEROKU_USE_SYSTEM_RUBY`, but I bet most buildpacks that choose to set this would just set it to 1 out of laziness. Though maybe not? We could get wild and enforce that it's semi-structured data like `buildpack: <name-of-buildpack>` or `os: <name-of-os-expected-to-have-ruby>`. I'm just trying to think of how we could give users more of a hint of where to look next when things go wrong other then "stuff failed"

It's useful to think about this now, as it would be a breaking change if we mandate something explicit/different in the future.